scipy/optimize/__minpack.h

```diff
@@ -672,37 +672,36 @@ static PyObject *minpack_lmder(PyObject *dummy, PyObject *args) {
 
   STORE_VARS();
 
-  if (!PyArg_ParseTuple(args, "OOO|OiidddidO", &fcn, &Dfun, &x0, &extra_args, &full_output, &col_deriv, &ftol, &xtol, &gtol, &maxfev, &factor, &o_diag)) return NULL;
+  if (!PyArg_ParseTuple(args, "OOiiO|OiidddidO", &fcn, &Dfun, &m_int, &n_int, &x0, &extra_args, &full_output, &col_deriv, &ftol, &xtol, &gtol, &maxfev, &factor, &o_diag)) return NULL;
+
+  n = n_int;
+  m = m_int;
 
   INIT_JAC_FUNC(fcn,Dfun,extra_args,col_deriv,minpack_error);
 
   /* Initial input vector */
-  ap_x = (PyArrayObject *)PyArray_ContiguousFromObject(x0, NPY_DOUBLE, 1, 1);
+  ap_x = (PyArrayObject *)PyArray_FromAny(x0, PyArray_DescrFromType(NPY_DOUBLE), 1, 1, NPY_ARRAY_DEFAULT | NPY_ARRAY_ENSUREARRAY | NPY_ARRAY_ENSURECOPY, NULL);
   if (ap_x == NULL) goto fail;
   x = (double *) PyArray_DATA(ap_x);
-  n = PyArray_DIMS(ap_x)[0];
+  if (n != PyArray_DIMS(ap_x)[0]) goto fail;
 
   if (maxfev < 0) maxfev = 100*(n+1);
 
-  /* Setup array to hold the function evaluations */
-  ap_fvec = (PyArrayObject *)call_python_function(fcn, n, x, extra_args, 1, minpack_error, -1);
-  if (ap_fvec == NULL) goto fail;
-  fvec = (double *) PyArray_DATA(ap_fvec);
-
   SET_DIAG(ap_diag,o_diag,mode);
 
-  m = (PyArray_NDIM(ap_fvec) > 0 ? PyArray_DIMS(ap_fvec)[0] : 1);
-
   dims[0] = n; dims[1] = m;
   ap_ipvt = (PyArrayObject *)PyArray_SimpleNew(1,&n,NPY_INT);
   ap_qtf = (PyArrayObject *)PyArray_SimpleNew(1,&n,NPY_DOUBLE);
   ap_fjac = (PyArrayObject *)PyArray_SimpleNew(2,dims,NPY_DOUBLE);
+  /* Setup array to hold the function evaluations */
+  ap_fvec = (PyArrayObject *)PyArray_SimpleNew(1,&m,NPY_DOUBLE);
```

Comment:
doesn't this have to be cleaned up at the end, i.e. `Py_DECREF`/`Py_XDECREF`?

Reply:
It does. For that reason, I left the calls to

```
Py_DECREF(ap_fvec);
```